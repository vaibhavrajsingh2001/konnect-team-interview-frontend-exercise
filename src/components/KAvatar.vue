<template>
  <img
    :alt="name"
    class="k-avatar"
    :src="avatarUrl"
    :title="title"
  >
</template>
<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'KAvatar',
  props: {
    avatarUrl: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
    email: {
      type: String,
      required: true,
    },
  },
  computed: {
    title(): string {
      return `${this.name} <${this.email}>`
    },
  },
})
</script>
<style lang="scss">
.k-avatar {
  height: 3.6rem;
  width: 3.6rem;
  border-radius: 50%;
}
</style>
