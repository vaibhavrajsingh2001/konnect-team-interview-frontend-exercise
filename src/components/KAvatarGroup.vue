<template>
  <div class="k-avatar-group">
    <k-avatar
      v-for="item in slicedAvatarList"
      :key="item.id"
      :avatar-url="item.avatar"
      :email="item.email"
      :name="item.name"
    />

    <!-- Element to show extra avatar count -->
    <span
      v-if="extraAvatarCount > 0"
      class="k-avatar-extra"
      :title="extraAvatarTitle"
    >
      +{{ extraAvatarCount }}
    </span>
  </div>
</template>
<script lang="ts">
import { defineComponent, type PropType } from 'vue'

import KAvatar from '@/components/KAvatar.vue'
import type { Developer } from '@/types'

export default defineComponent({
  name: 'KAvatarGroup',
  components: {
    KAvatar,
  },
  props: {
    avatarList: {
      type: Array as PropType<Developer[]>,
      required: true,
    },
  },
  computed: {
    maxAvatarCount(): number {
      return 2
    },
    slicedAvatarList(): Developer[] {
      return this.avatarList.slice(0, this.maxAvatarCount)
    },
    extraAvatarCount(): number {
      return this.avatarList.length - this.maxAvatarCount
    },
    extraAvatarTitle(): string {
      let titleString = ''
      const avatarsNotShown = this.avatarList.slice(this.maxAvatarCount)

      avatarsNotShown.forEach((avatar, index) => {
        // Add a new line after each avatar except the last one
        titleString += `${avatar.name} <${avatar.email}> ${index < (avatarsNotShown.length - 1) ? '\n\n' : ''}`
      })

      return titleString
    },
  },
})
</script>
<style lang="scss">
.k-avatar-group {
  display: flex;
}

// For overlapping avatars:
// Add negative margin to all child elements of the group, except the first child
.k-avatar-group>*:not(:first-child) {
  margin-left: -1rem;
}

.k-avatar-extra {
  align-items: center;

  background-color: #F1F1F8;
  border: 1px solid #fff;

  border-radius: 50%;
  color: #777D8A;

  cursor: default;

  display: inline-flex;
  font-size: 1.2rem;
  font-weight: 600;
  height: 3.6rem;
  justify-content: center;
  letter-spacing: 0.2rem;
  width: 3.6rem;
}
</style>
