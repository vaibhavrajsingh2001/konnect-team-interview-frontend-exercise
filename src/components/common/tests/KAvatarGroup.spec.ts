import { renderToString, mount } from '@vue/test-utils'
import { it, describe, expect } from 'vitest'
import KAvatarGroup from '@/components/common/KAvatarGroup.vue'
import KAvatar from '@/components/common/KAvatar.vue'

const sampleDevList = [
  { id: '62b2988c-4178-48cb-8343-eb15f7924a84', name: 'Timmy Rosenbaum', email: 'Timmy_Rosenbaum@hotmail.com', avatar: 'https://avatars.githubusercontent.com/u/16572492' },
  { id: '3a460c26-0dca-45d8-9d2a-a94202fba68a', name: 'Theodore Aufderhar', email: 'Theodore_Aufderhar@hotmail.com', avatar: 'https://avatars.githubusercontent.com/u/64255661' },
  { id: 'ab49cfc6-04d6-44e8-8d96-f3d7faa2f0a8', name: 'Tanya Ferry', email: 'Tanya.Ferry47@yahoo.com', avatar: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/801.jpg' },
  { id: 'e189c250-3eeb-494e-a8ca-31d4b9c23b33', name: 'Clayton Bins', email: 'Clayton_Bins30@gmail.com', avatar: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1126.jpg' },
]
describe('KAvatarGroup', () => {
  it('renders correctly for 2 developers', async () => {
    const wrapper = await renderToString(KAvatarGroup, {
      props: {
        avatarList: sampleDevList.slice(0, 2),
      },
      components: {
        KAvatar,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders extra avatar correctly', async () => {
    const wrapper = mount(KAvatarGroup, {
      props: {
        avatarList: sampleDevList,
      },
      components: {
        KAvatar,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('shows correct extra avatar count', async () => {
    const wrapper = mount(KAvatarGroup, {
      props: {
        avatarList: sampleDevList,
      },
      components: {
        KAvatar,
      },
    })

    expect(wrapper.findTestId('k-extra-avatar-count').isVisible()).toBe(true)
    expect(wrapper.findTestId('k-extra-avatar-count').text()).toBe('+2')
  })
})
