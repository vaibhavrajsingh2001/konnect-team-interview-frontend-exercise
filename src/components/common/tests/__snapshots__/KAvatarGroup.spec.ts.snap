// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KAvatarGroup > renders correctly for 2 developers 1`] = `"<div class="k-avatar-group"><!--[--><img alt="Timmy Rosenbaum" class="k-avatar" src="https://avatars.githubusercontent.com/u/16572492" title="Timmy Rosenbaum &lt;Timmy_Rosenbaum@hotmail.com&gt;"><img alt="Theodore Aufderhar" class="k-avatar" src="https://avatars.githubusercontent.com/u/64255661" title="Theodore Aufderhar &lt;Theodore_Aufderhar@hotmail.com&gt;"><!--]--><!-- Element to show extra avatar count --><!--v-if--></div>"`;

exports[`KAvatarGroup > renders extra avatar correctly 1`] = `
"<div class="k-avatar-group"><img alt="Timmy Rosenbaum" class="k-avatar" src="https://avatars.githubusercontent.com/u/16572492" title="Timmy Rosenbaum <Timmy_Rosenbaum@hotmail.com>"><img alt="Theodore Aufderhar" class="k-avatar" src="https://avatars.githubusercontent.com/u/64255661" title="Theodore Aufderhar <Theodore_Aufderhar@hotmail.com>"><!-- Element to show extra avatar count --><span class="k-avatar-extra" data-testid="k-extra-avatar-count" title="Tanya Ferry <Tanya.Ferry47@yahoo.com> 

Clayton Bins <Clayton_Bins30@gmail.com> "> +2</span></div>"
`;
